pkg/ui/src/views/jobs/jobDetails.tsx

```diff
@@ -0,0 +1,200 @@
+  is(...statuses: string[]) {
+    return statuses.indexOf(this.props.job.status) !== -1;
+  }
+
+  refresh = (props = this.props) => {
+    props.refreshJobs(new JobsRequest({
+      status: props.status,
+      type: props.type,
+      limit: parseInt(props.show, 10),
+    }));
+  }
+
+  componentWillMount() {
+    this.refresh();
+  }
+
+  prevPage = () => this.props.history.goBack();
+
+  renderProgress() {
+    if (this.is(JOB_STATUS_SUCCEEDED, JOB_STATUS_FAILED, JOB_STATUS_CANCELED)) {
+      const className = classNames("jobs-table__status", {
+        "jobs-table__status--succeed": this.is(JOB_STATUS_SUCCEEDED),
+        "jobs-table__status--failed": this.is(JOB_STATUS_FAILED, JOB_STATUS_CANCELED),
+      });
+      return (
+        <span className={className}>
+          {renamedStatuses(this.props.job.status)}
+        </span>
+      );
+    }
+    const percent = this.props.job.fraction_completed * 100;
+    return (
+      <div className="jobs-table__progress">
+        {this.props.job.running_status
+          ? <div className="jobs-table__running-status">{this.props.job.running_status}</div>
+          : null}
+
+        <Line
+          percent={percent}
+          strokeWidth={2}
+          trailWidth={2}
+          strokeColor="#0788ff"
+          trailColor="#d6dbe7"
+          className="jobs-table__progress-bar"
+        />
+      </div>
+    );
+  }
+
+  renderDuration() {
+    const { job } = this.props;
+    const started = TimestampToMoment(job.started);
+    const finished = TimestampToMoment(job.finished);
+    const modified = TimestampToMoment(job.modified);
+    if (this.is(JOB_STATUS_PENDING, JOB_STATUS_PAUSED)) {
+      return _.capitalize(this.props.job.status);
+    } else if (this.is(JOB_STATUS_RUNNING)) {
+      const fractionCompleted = this.props.job.fraction_completed;
+      if (fractionCompleted > 0) {
+        const duration = modified.diff(started);
+        const remaining = duration / fractionCompleted - duration;
+        return <span className="jobs-table__duration--right">{formatDuration(moment.duration(remaining)) + " remaining"}</span>;
+      }
+    } else if (this.is(JOB_STATUS_SUCCEEDED)) {
+      return <span>{"Duration: " + formatDuration(moment.duration(finished.diff(started)))}</span>;
+    }
+  }
+
+  renderStatus = () => {
+    const { job } = this.props;
+    const percent = job.fraction_completed * 100;
+    switch (job.status) {
+      case JOB_STATUS_SUCCEEDED:
+        return <div className="job-status__line">{this.renderProgress()} - {this.renderDuration()}</div>;
+      case JOB_STATUS_FAILED || JOB_STATUS_CANCELED:
+        return <div>{this.renderProgress()}</div>;
+      default:
+        return <div>
+          {this.renderProgress()}
+          <div className="job-status__line--percentage"><span>{percent.toFixed() + "%"} done</span><Divider type="vertical" />{this.renderDuration()}</div>
+        </div>;
+    }
+  }
+
+  renderContent = () => {
```

Comment:
Components with a lot of nested `.renderX()` calls are difficult to read and debug. Can you extract into separate components with props?

I would love to see component names clarify the render hierarchy instead of `this.renderBlah`.

For example, try simple presentational components like `<JobStatus job={job} />`. Then we can introduce much easier testing of the low level components in enzyme. Right now, the `JobDetails` component is very tricky to test because it requires too many inputs.

Reply:
Components will be extracted.